### DISCUSSION TITLE
How to properly detect who is calling service

### DISCUSSION THREAD

**Discussion by bjornbouetsmith** [CoreWCF/CoreWCF#1330] [Q&A]
Hi,

We are in the process of converting our NET framework WCF Services to NET6 and CoreWCF - and so far smooth sailing - but many times every day we get the following errors logged in our error logs:

`11:06:24.3551|ERROR|Microsoft.AspNetCore.Server.Kestrel|Unhandled exception while processing 0HN1291DG6PFE.|CoreWCF.IdentityModel.Tokens.SecurityTokenValidationException: The service does not allow you to log on anonymously.
   at CoreWCF.Security.SecurityUtils.ValidateAnonymityConstraint(WindowsIdentity identity, Boolean allowUnauthenticatedCallers)
   at CoreWCF.Channels.WindowsStreamSecurityUpgradeProvider.WindowsStreamSecurityUpgradeAcceptor.CreateClientSecurityAsync(NegotiateStream negotiateStream, Boolean extractGroupsForWindowsAccounts)
   at CoreWCF.Channels.WindowsStreamSecurityUpgradeProvider.WindowsStreamSecurityUpgradeAcceptor.OnAcceptUpgradeAsync(Stream stream)
   at CoreWCF.Channels.StreamSecurityUpgradeAcceptorBase.AcceptUpgradeAsync(Stream stream)
   at CoreWCF.Channels.Framing.ServerFramingDuplexSessionMiddleware.UpgradeConnectionAsync(FramingConnection connection, String contentType)
   at CoreWCF.Channels.Framing.ServerFramingDuplexSessionMiddleware.OnConnectedAsync(FramingConnection connection)
   at CoreWCF.Channels.Framing.NetMessageFramingConnectionHandler.PerformServiceHandshake(IFramingConnectionHandshakeBuilder configuration, FramingConnection connection, HandshakeDelegate next)
   at CoreWCF.Channels.Framing.DuplexFramingMiddleware.OnConnectedAsync(FramingConnection connection)
   at CoreWCF.Configuration.MapMiddleware.Invoke(FramingConnection connection)
   at CoreWCF.Channels.Framing.FramingModeHandshakeMiddleware.OnConnectedCoreAsync(FramingConnection connection)
   at CoreWCF.Channels.Framing.FramingModeHandshakeMiddleware.OnConnectedAsync(FramingConnection connection)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.KestrelConnection´1.ExecuteAsync()`

We already have an exception handler injected to handle exceptions from the code, so all unhandled exceptions/thrown exceptions are handled correctly and we also have an operation invoker in place for our custom logging.

But I would like to know if its possible to intercept these failed requests and log information about what request caused this.

We do not allow anonymous requests for most of our services.

To me this error seems to be coming too early for any of our code to be able to intercept, but perhaps we are just missing something?

I have tried to see if I could reproduce it with a test, where I create a service with 
` new System.ServiceModel.NetTcpSecurity
  {
      Mode = System.ServiceModel.SecurityMode.Transport,
      Transport =
  {
      ClientCredentialType = System.ServiceModel.TcpClientCredentialType.Windows,
      ProtectionLevel = ProtectionLevel.None
  },
  };`

And a client with no security at all, but client just immediately faults and nothing is logged server side.

Thanks in advance for any hints.

**Comment by mconnew** [comment]
What information would you like to see? Are you looking for the client IP address? Client IP and port number? I suspect the client is doing something the equivalent of `WindowsIdentity.GetAnonymous()` and impersonating that user. They might not be using that exact code, but the situation is likely equivalent.

**Reply by bjornbouetsmith** [reply]
It would be nice to see what service/method called - and the remote client address?

I guess many people run several services in same host process, and without knowing service/method called it makes it hard to diagnose.

Without client IP, almost impossible, since you would need to either turn off tracing of calls (if possible in CoreWCF) - or start up wireshark and hope that a call is being done that does this.

`Unhandled exception while processing 0HN1291DG6PFE.` is not a lot of information - but I would also be happy enough if this error was suppressed and only went into trace logs similar to what NET Framework would most likely do.

**Reply by mconnew** [reply]
Sorry, missed your response. I'm trying to get better as using GitHub notifications, it's a process. Anyway, you can find the client's ip by getting the `RemoteEndpointMessageProperty` from the incoming message properties. You can do something like this:
```c#
string? address = null;
int port;
if (OperationContext.Current.IncomingMessageProperties.TryGetValue<RemoteEndpointMessageProperty>(RemoteEndpointMessageProperty.Name, out RemoteEndpointMessageProperty remp))
{
    address = remp.Address;
    port = remp.Port;
}
```

In the next release we're adding injection support for properties which will make this even easier. You will be able to do this:
```c#
[ServiceContract]
public interface IService
{
    [OperationContract]
    void DoSomething(string data);
}

public partial class Service : IService
{
    public void DoSomething(string data, [CoreWCF.Injected(PropertyName = CoreWCF.Channels.RemoteEndpointMessageProperty.Name)] CoreWCF.Channels.RemoteEndpointMessageProperty remp)
    {
        string address = remp.Address;
        int port = remp.Port;
    }
}
```

That's not released yet, but very soon.  

If you have some helper class you are calling which needs to log this so is unaware of what method/action was called, you can get that info from the incoming message headers.
```c#
string action = OperationContext.Current.IncomingMessageHeaders.Action;
```
From the action you can map that to an interface method. If you have an app with multiple services implementing the same interface and need to know which service type the call was made against, you can do this:
```c#
Type serviceType = OperationContext.Current.Host.Description.ServiceType;
```
If you want to know which Uri the request was received on (same contract, multiple endpoints, eg with different bindings), you can find that like this:
```c#
Uri receivedUri = OperationContext.Current.IncomingMessageHeaders.To;
```

Hopefully that gives you everything you need. Let me know if there's more info you would need.

**Reply by bjornbouetsmith** [reply]
@mconnew We already use the message headers etc, but the problem is that the error message that I pasted in gets emitted before it hits any of our own implemented code.

Its the CoreWCF code that throws this exception and that exception does not bubble up to our code, so we are in the dark here.

The only reason we get this error logged is because our logging framework have injected something into the error logging pipeline in asp.net - and my guess is that the OperationContext is gone.

So I guess the only way to "get it correctly logged" - is to inject another logging framework that extract these headers and hope that they are available at the time the log even is raised.

**Reply by mconnew** [reply]
How do you handle this in WCF on .NET Framework? Whatever mechanism you use there, we should make sure works in CoreWCF too. If this is something you're doing extra over what you did with WCF, I did technically make the the NetFraming stack extendable with middleware similar to aspnetcore middleware with the method signature being `OnConnectedAsync(FramingConnection connection)`. If it turns out that's what's needed, I can put together a quick example showing how to extend that.

**Reply by bjornbouetsmith** [reply]
> How do you handle this in WCF on .NET Framework? Whatever mechanism you use there, we should make sure works in CoreWCF too. If this is something you're doing extra over what you did with WCF, I did technically make the the NetFraming stack extendable with middleware similar to aspnetcore middleware with the method signature being `OnConnectedAsync(FramingConnection connection)`. If it turns out that's what's needed, I can put together a quick example showing how to extend that.

@mconnew  We dont, and we live in ignorant bliss - since the WCF framework drops the messages before it hit any user code - and only with diagnostics tracing via .svclogs is this available. But having this available would be a very nice improvement, so its easier to detect and correct these wrong connections - only reason we noticed this is because asp.net is emitting these exceptions that gets logged by the logging framework. In old WCF these would be "hidden"

**Reply by onidio** [reply]
@mconnew
> How do you handle this in WCF on .NET Framework? Whatever mechanism you use there, we should make sure works in CoreWCF too. If this is something you're doing extra over what you did with WCF, I did technically make the the NetFraming stack extendable with middleware similar to aspnetcore middleware with the method signature being `OnConnectedAsync(FramingConnection connection)`. If it turns out that's what's needed, I can put together a quick example showing how to extend that.

If the OnConnectedAsync(FramingConnection) solution already exists, I'd be interested in a quick code example

**Comment by onidio** [comment]
Hello, we are facing exactly the same problem. Have you found a solution?

**Reply by mconnew** [reply]
Hopefully my answer [here should help](https://github.com/CoreWCF/CoreWCF/discussions/1330#discussioncomment-10183268). Let me know if there's anything else you need. Mention me with an @ to make it easier for me to see your response.

**Comment by onidio** [comment]
@mconnew
1) CoreWcf not working like Old Wcf!
The first thing I noticed (as did @bjornbouetsmith) is that, unlike with Legacy Wcf, authentication errors are triggered far too top and are therefore not intercepted by our IErrorHandler ServiceBehavior. To reproduce the problem, all you have to do, as in our case, is use a CustomCertificateValidator to validate certificates: with the old WCF, if an exception occurred in the Validate procedure, it was redirected to our IErrorHandler ServiceBehavior, but this is unfortunately no longer the case with CoreWcf.

2) Desired improvement 
Using the CustomCertificateValidator and the IErrorHandler ServiceBehavior enabled us to log problems with invalid certificates, but we were still unable to obtain details of the client causing the error (see https://github.com/CoreWCF/CoreWCF/discussions/1416); this is why I would have liked to be able to access OperationContext.Current in the CustomCertificateValidator and the IErrorHandler ServiceBehavior, which unfortunately is still null in these two contexts.

We're in the middle of evaluating the possibility of migrating to CoreWcf, and everything has gone smoothly (and I congratulate you on the work you've done), so the only remaining concern is the two points above.

**Reply by mconnew** [reply]
I took a look at the code in WCF and I could only find usage of IErrorHandler from the dispatcher layer. For NetTcp, the certificate validator is called as part of the connection handshake when a channel is opened and I couldn't find a call into IErrorHandler there. I'm not saying it doesn't do that, just that I couldn't find the code which does it so I could make sure equivalent code exists in CoreWCF.  Can you replicate the failure scenario and capture the call stack when your IErrorHandler is invoked, that way I can more easily work out what the difference is.

**Reply by mconnew** [reply]
I also wrote up a description of how you can achieve this today with a bunch of boiler plate code in the other discussion you created if you can't wait to resolve this in a way similar to how WCF does things.

**Reply by bjornbouetsmith** [reply]
> I took a look at the code in WCF and I could only find usage of IErrorHandler from the dispatcher layer. 
in the old WCF code, everything in out code is handled by an IOperationInvoker implementation, and we just wrap all calls in that.
Obviously that does not catch calls that does not even hit the service layer, i.e. stuff that gets rejected because of contract mismatches etc. 
But for normal exception handling stuff this is perfect.

So if we can extend CoreWCF to make this even better, so an actual IErrorHandler implementation is atually invoked at different points in the pipeline, i.e. as part of connection establishment, contract resolving etc. then it would be amazing, since then we can much better log server side if some "rogue" client is calling the wrong endpoints and is getting rejected.

In old WCF, an error like the one above would not even be caught by our logging, because it would be silently rejected by WCF and only if you enabled diagnostics tracing to .svclog files you would find these calls - but extending corewcf so its "easier" to decide whether or not to log these in code and not configuration would be amazing

**Reply by onidio** [reply]
@mconnew
> I took a look at the code in WCF and I could only find usage of IErrorHandler from the dispatcher layer. For NetTcp, the certificate validator is called as part of the connection handshake when a channel is opened and I couldn't find a call into IErrorHandler there. I'm not saying it doesn't do that, just that I couldn't find the code which does it so I could make sure equivalent code exists in CoreWCF. Can you replicate the failure scenario and capture the call stack when your IErrorHandler is invoked, that way I can more easily work out what the difference is.

Here's a screenshot of the call stack in the case of an exception in the Validate procedure of a CustomCertificatValidator in WCF
![vVpqBbKCIT](https://github.com/user-attachments/assets/5d79c020-cd67-4c74-a4e8-e2a1b66f9491)
